### DISCUSSION TITLE
Use in async and/or tokio

### DISCUSSION THREAD

**Discussion by computerquip** [jam1garner/binrw#178] [Q&A]
Is there any way to use this functionality using tokio in an asynchronous manner?

**Comment by csnover** [comment]
It should be possible to use binrw with tokio by using [`tokio::task::spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) or similar.

**Comment by dizda** [comment]
There's no way to read/write asynchronously?

**Reply by kitlith** [reply]
Not at the current moment. That's waiting on either someone who wants to implement it with `#[async_trait]`, or language support for async in traits.

(aside, I'm not sure if there's a good way to turn either async functions or sync functions into eachother, which means that if binrw tried to be both, manual trait implementations would essentially have to be written twice. guess we have another argument for effect generics >_>)

**Reply by dizda** [reply]
Thank you for the honest response.

**Comment by csnover** [comment]
For anyone who is interested in a more detailed write-up on async library support in Rust, the blog post [The bane of my existence: Supporting both async and sync code in Rust](https://nullderef.com/blog/rust-async-sync/) provides more information about why it is not so simple right now.
